server/src/main/java/org/opensearch/index/codec/CodecType.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.codec;
+
+import org.apache.lucene.codecs.Codec;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.index.engine.EngineConfig;
+
+import java.util.Set;
+
+/**
+ * A codec enum that maintains a list of compression codecs alongside their respective supported index settings.
+ *
+ * @opensearch.internal
+ */
+public enum CodecType {
```

Comment:
I like the idea of codec-related metadata, but I think the the enumeration is not the right choice there: codecs are extensible (as well as `CodecService`), just look at the way ZSTD was introduced initially using the plugins. With enumeration we are artificially constraining the list of the codecs to what we support in core, leaving no room for plugin developers to add new codecs easily.

I would suggest to follow a different implementation and explicitly manage compression settings in `CodecService`:

```
public class CodecService {
    public int compressionLevel(String codecName, IndexSettings indexSettings) {
        ....
    } 
}
```

The implementors would be able to override the function and add compression levels for their codecs.

Reply:
The only catch is every new setting that is getting added for codecs would require a new method with a default implementation to be introduced in CodecService which should be okay.

Reply:
Agree, I think this is fine since the semantic of the setting is very precise

Reply:
We don't have a CodecService instance during validation for compression level settings, so we might have to put in a static method which does the validation. 
To fetch compression level, we can have a separate method

Reply:
> We don't have a CodecService instance during validation for compression level settings,

What do you mean by that?

Reply:
> > We don't have a CodecService instance during validation for compression level settings,
> 
> What do you mean by that?

I was just pointing out that [here](https://github.com/opensearch-project/OpenSearch/pull/8737/files#diff-2619776bfaaef5455bdd5fab7b3b5a4811233f399fb4905b8de32b7ec2e54589R181) we are performing the validation if the compression_level is being set for the right codec or not, but we do not have a instance of CodecService over here to check via something like this ```public int compressionLevel(String codecName, IndexSettings indexSettings)```
We won't be able to keep the validations inside the CodecService because we would want to fail the index update settings request, which happens when the index is closed, so we introduced SettingsDependency.

Is my understanding based on the above discussion correct?